src/lib.rs

```diff
@@ -42,22 +60,63 @@ impl TsnSocket {
 
 const CONTROL_SOCK_PATH: &str = "/var/run/tsn.sock";
 
-fn send_cmd(command: String) -> Result<String, std::io::Error> {
-    let mut stream = UnixStream::connect(CONTROL_SOCK_PATH)?;
-    stream.write_all(command.as_bytes())?;
-    let mut msg = String::new();
-    stream.read_to_string(&mut msg)?;
-    Ok(msg)
+#[derive(Eq, PartialEq)]
+enum LockKind {
+    LOCK,
+    UNLOCK,
 }
 
-fn create_vlan(ifname: &str, vlanid: u16) -> Result<String, std::io::Error> {
-    let command = format!("create {} {}\n", ifname, vlanid);
-    send_cmd(command)
+fn create_vlan(ifname: &str, vlanid: u16) -> Result<i32, i32> {
+    let configs = config::read_config(CONFIG_PATH).unwrap();
+    let config = configs.get(ifname).unwrap();
+    let shm_name = format!("{}.{}", ifname, vlanid);
+    let shm_fd = shm_open(
+        &*shm_name,
+        OFlag::O_CREAT | OFlag::O_RDWR,
+        Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+    )
+    .unwrap();
+    lock_shmem(&shm_fd, LockKind::LOCK);
+    let mut vlan_vec = read_shmem(&*shm_name);
+    let mut result = Ok(0);
+    if vlan_vec.len() == 0 {
+        result = vlan::create_vlan(config, ifname, vlanid);
+    }
+    vlan_vec.push(process::id());
+    write_shmem(&*shm_name, &vlan_vec);
+    lock_shmem(&shm_fd, LockKind::UNLOCK);
+    result
 }
 
-fn delete_vlan(ifname: &str, vlanid: u16) -> Result<String, std::io::Error> {
-    let command = format!("delete {} {}\n", ifname, vlanid);
-    send_cmd(command)
+fn delete_vlan(ifname: &str, vlanid: u16) -> Result<i32, i32> {
+    let shm_name = format!("{}.{}", ifname, vlanid);
```

Comment:
Extract shmem name as a function to avoid use different name between create and delete